execute/observation.go

```diff
@@ -259,68 +261,177 @@ func buildCombinedReports(
+	msgHashes := make(exectypes.MessageHashes)
 
-	commitObs := regroup(commitData)
+	sort.Slice(commitData, func(i, j int) bool {
+		return exectypes.CompareCommitData(commitData[i], commitData[j])
+	})
 
-	for srcChain, reports := range commitObs {
-		if len(reports) == 0 {
-			continue
+	stop := false
+	gasSum := uint64(0)
+
+	for _, report := range commitData {
+		srcChain := report.SourceChain
+
+		// Read messages for this report's sequence number range
+		msgs, err := p.readMessagesForReport(ctx, lggr, srcChain, report)
+		if err != nil {
+			continue // Logging already done in readMessagesForReport
 		}
 
-		messageObs[srcChain] = make(map[cciptypes.SeqNum]cciptypes.Message)
-		// Read messages for each range.
-		for _, report := range reports {
-			msgs, err := ccipReader.MsgsBetweenSeqNums(ctx, srcChain, report.SequenceNumberRange)
-			if err != nil {
-				lggr.Errorw("unable to read all messages for report",
-					"srcChain", srcChain,
-					"seqRange", report.SequenceNumberRange,
-					"merkleRoot", report.MerkleRoot,
-					"err", err,
-				)
-				continue
+		// Add the report to available reports
+		availableReports[srcChain] = append(availableReports[srcChain], report)
+
+		// Initialize data structures for this source chain if needed
+		initSourceChainMaps(srcChain, messageObs, msgHashes)
+		// Process each message in the report
+		for _, msg := range msgs {
+			seqNum := msg.Header.SequenceNumber
+
+			// Handle message observation and gas calculation
+			if !stop && !p.inflightMessageCache.IsInflight(srcChain, msg.Header.MessageID) {
+				messageObs[srcChain][seqNum] = msg
+				gasSum += p.estimateProvider.CalculateMessageMaxGas(msg)
+				stop = p.exceedsMaxGasLimit(gasSum, msgs)
+			} else {
+				// This is for when we calculate roots in reports later, we need the seqNum and
+				// the hash for this message
+				messageObs[srcChain][seqNum] = createEmptyMessageWithSeqNum(seqNum)
 			}
 
-			if !msgsConformToSeqRange(msgs, report.SequenceNumberRange) {
-				lggr.Errorw("missing messages in range",
-					"srcChain", srcChain, "seqRange", report.SequenceNumberRange)
-				continue
+			// Compute hash for all messages in report even if they are executed or not included in this round
+			hash, err := p.msgHasher.Hash(ctx, msg)
+			if err != nil {
+				return exectypes.Observation{}, fmt.Errorf("unable to hash message srcChain %d, seqNum %d: %w",
+					srcChain, seqNum,
+					err)
 			}
+			msgHashes[srcChain][seqNum] = hash
 
-			for _, msg := range msgs {
-				messageObs[srcChain][msg.Header.SequenceNumber] = msg
-				messageTimestamps[msg.Header.MessageID] = report.Timestamp
+			// Update the observation with current state
+			observation.CommitReports = availableReports
+			observation.Messages = messageObs
+			observation.Hashes = msgHashes
+
+			// Check if we've exceeded encoding size limits
+			if !stop {
+				stop = p.handleEncodingSizeLimits(observation, messageObs, srcChain, seqNum)
```

Comment:
Having `handleEncodingSizeLimits` be a receiver function makes me a little nervous that there could be a side effect. Could you make it a pure function?

Reply:
If this message causes the limit to be hit, should it be removed?
```suggestion
				stop = p.handleEncodingSizeLimits(observation, messageObs, srcChain, seqNum)
				if stop {
					messageObs[srcChain][seqNum] = createEmptyMessageWithSeqNum(seqNum)
				}
```

Reply:
That's why I'm using lenient maxSize, The max size is set up to use 75% of actual max that we're looking for to allow for these things

Reply:
I think it's better to be consistent. Room is still needed for the remaining hashes. The worst case scenario is a commit report with 256 messages and we hit the limit on message 2. If message 2 is huge will there always be enough room for the remaining 254 hashes?

The builder also has logic to handle skipping the large message and continuing with smaller messages. So I wonder if it's possible that this optimization could cause problems with other optimizations in the builder.